Adds configurable behavior on redirect to loopback

d2r opened the pull request:
Requesting a backport to 8.x because although this has a functionality change with scope that is best reserved for major releases, 8.x is not yet released. It is intended to be considered for the 8.x release rather than waiting for another major release.

[Discussion Thread](https://lists.apache.org/thread.html/c46e9f0493896744ffbfb6eb0d8eed53bb77003a8abbfa1ac5b63c16@%3Cdev.trafficserver.apache.org%3E)

~~Depends on #4166 because a new assert gets tripped in this failure mode.~~
Dropping the dependency on #4166 because the bug was determined to be local to this PR and not directly related to #4166.

d2r commented on doc/admin-guide/files/records.config.en.rst:
```diff
@@ -1326,6 +1326,24 @@ HTTP Redirection
+.. warning:: Following a redirect to a loopback address can be dangerous, as it allows the controller of an origin to arrange a probe the Trafficserver host. Generally, an administrator should never configure Trafficserver to follow these redirects.
+
+The default behavior is to return the origin response to the client.
+
```
Note to reviewers:

<img width="996" alt="screen shot 2018-08-22 at 00 52 02" src="https://user-images.githubusercontent.com/905298/44445318-188f5180-a5a6-11e8-8bc2-c9987e553f56.png">

This is a rendering in my browser. The big idea was that an administrator who uses the dangerous setting is fully informed of the risk, and the reader knows the default behavior is reasonable.

d2r replied:
This config has been changed to `redirect_enabled` and the doc has been expanded and updated since this screenshot.

d2r commented on tests/gold_tests/redirect/redirect-deny-loopback.test.py:
```diff
@@ -0,0 +1,143 @@
+# We can add them here for manual testing.
+for t in (['127.0.0.1']):
+    for s in ScenarioE:
+        makeTestCase(redirectTarget=t, scenario=s)
```
For automated tests that gate things like PR acceptance, this should be sensible.

However, a more complete test would include a collection of different ways to address a loopback device. This is difficult to accommodate at the level where the test runs, as we would need infrastructure in the loop to arrange for interfaces at such addresses.

So that is why I have left the outer loop here with comments to the effect.

d2r replied:
This is now picking all the non-link-local IPv6 as well as IPv4 addresses of the host.

d2r commented on proxy/http/HttpTransact.cc:
```diff
@@ -1546,6 +1546,28 @@ HttpTransact::OSDNSLookup(State *s)
+        build_response_copy(s, &s->hdr_info.server_response, &s->hdr_info.client_response, s->client_info.http_version);
+        TRANSACT_RETURN(SM_ACTION_INTERNAL_CACHE_NOOP, nullptr);
+    }
+  }
```
Note to reviewers:

This follows the lazy consensus from the [discussion](https://lists.apache.org/thread.html/c46e9f0493896744ffbfb6eb0d8eed53bb77003a8abbfa1ac5b63c16@%3Cdev.trafficserver.apache.org%3E) on the dev list.

We were sure we wanted this to be configurable, because there are several plausible alternatives.
We were sure the default should not be to follow the redirect.

When the admin configures the proxy to follow up to *N* redirects, if the *N+1*st redirect is to loopback it is not clear what would be expected. Should we simply return the 3xx as-is, or should we stop and inspect its `Location` header to see whether it is loopback and reject it if so?

Since there seemed to be this little ambiguity, I chose the default behavior to simply return the 3xx response to the client. (One can still configure the 403.)

d2r replied:
Updated as per discussion.

d2r commented:
Labeled WIP so I don't forget to squash commits.

bryancall commented:
You might want to add another option to add non-routable IPs to the list too.  I would create a configuration option called proxy.config.http.redirect_enabled (with all the options) and start with 0 as off and set proxy.config.http.number_of_redirections to 1 as the default.

When checking for localhost you should resolve the ip and check to see if it matches any of the IPs on the server.  That could be a long list if there servers has many IP aliases.  This still doesn’t block IPs used by other servers in the same network.

d2r commented:
OK, taking discussion back to the [mail thread](https://lists.apache.org/thread.html/c46e9f0493896744ffbfb6eb0d8eed53bb77003a8abbfa1ac5b63c16@%3Cdev.trafficserver.apache.org%3E) so we can converge on how we want to do this.

d2r commented on doc/admin-guide/files/records.config.en.rst:
```diff
@@ -1300,6 +1300,34 @@ HTTP Connection Timeouts
+.. warning:: Following a redirect to an abnormal address can be dangerous, as it allows the controller of an origin to arrange a probe the Trafficserver host. Generally, an administrator should never configure Trafficserver to follow these redirects.
+
+The default behavior is to return the origin response to the client.
+
```
<img width="998" alt="screen shot 2018-08-23 at 23 27 13" src="https://user-images.githubusercontent.com/905298/44565115-21119480-a72c-11e8-9213-514833c049e9.png">

A rendering at commit 0d42475667.

d2r commented on iocore/utils/Machine.cc:
```diff
@@ -288,6 +288,19 @@ Machine::is_self(const IpAddr *ipaddr)
+  }
+  ats_ip_ntop(addr, string_value, sizeof(string_value));
+  return ink_hash_table_lookup(machine_id_ipaddrs, string_value, &value) == 1 ? true : false;
+}
```
NOTE: This is on the critical path, so I added this to avoid needing to construct an `IpAddr`.

d2r commented on proxy/http/HttpTransact.cc:
```diff
@@ -1546,6 +1546,30 @@ HttpTransact::OSDNSLookup(State *s)
 
+  if (s->redirect_info.redirect_in_process) {
+    if (HttpConfigParams::REDIRECT_ENABLED_FOLLOW == s->http_config_param->redirect_enabled) {
+      TxnDebug("http_trans", "[OSDNSLookup] Invalid redirect address. Following");
```
Since we have 5 separate checks to do when not configured to follow, I am handling this case separately.

d2r commented on tests/gold_tests/redirect/redirect-handle-abnormal-address.test.py:
```diff
@@ -0,0 +1,165 @@
+ipv6addrs = ["[{0}]".format(ip.split('%')[0]) for \
+        (family,_,_,_,(ip,*_)) in \
+        socket.getaddrinfo(host,port=None) if \
+        socket.AF_INET6 == family and 'fe80' != ip[0:4]] # Skip link-local addresses.
```
NOTE: I recognize this test will do different things on differently-configured systems. But it should not error in any case. I had issues with link-local IPv6 addresses in the Follow scenario, so I have expressly omitted them. I am not sure how platform-independent such test cases would be.

d2r commented on tests/gold_tests/redirect/redirect-handle-abnormal-address.test.py:
```diff
@@ -0,0 +1,165 @@
+        '127.1.2.3', '[::1]',      # loopback
+        '169.254.0.1', '[fe80::]', # link-local
+        '224.1.2.3', '[ff42::]',   # multicast
+        '10.0.0.1', '[fc00::1]',   # private
```
NOTE: These are not exhaustive. I am not testing whether our detection of all the possible values is correct, only that we behave correctly when configured to detect such classes.

d2r commented:
Implementation and tests should be working. I am still looking at updating the document.

d2r commented:
[approve ci autest]

d2r commented on doc/admin-guide/files/records.config.en.rst:
```diff
@@ -1300,6 +1300,8 @@ HTTP Connection Timeouts
 ================
 
+The default behavior is to return the origin response to the client.
+
```
Missed this in last check. Will remove.

d2r commented on proxy/http/HttpConfig.cc:
```diff
@@ -1487,6 +1488,8 @@ HttpConfig::reconfigure()
   params->oride.client_cert_filename        = ats_strdup(m_master.oride.client_cert_filename);
   params->oride.client_cert_filepath        = ats_strdup(m_master.oride.client_cert_filepath);
+  params->redirect_actions_string           = ats_strdup(m_master.redirect_actions_string);
+  params->redirect_actions_map = parse_redirect_actions(params->redirect_actions_string, params->redirect_actions_self_action);
```
`clang-format` made me do this.

d2r commented on proxy/http/HttpConfig.h:
```diff
@@ -395,6 +397,39 @@ OptionBitSet optStrToBitset(std::string_view optConfigStr, ts::FixedBufferWriter
+  {"default", AddressClass::DEFAULT},     {"private", AddressClass::PRIVATE},     {"loopback", AddressClass::LOOPBACK},
+  {"multicast", AddressClass::MULTICAST}, {"linklocal", AddressClass::LINKLOCAL}, {"routable", AddressClass::ROUTABLE},
+  {"self", AddressClass::SELF},
+};
```
This is what `clang-format` produced.

d2r commented:
[approve ci autest]

d2r commented on tests/gold_tests/redirect/redirect_actions.test.py:
```diff
@@ -0,0 +1,225 @@
+import socket
+Test.Summary = '''
+Test redirection behavior to invalid addresses
+'''
```
We may want to add:
```Python
Test.SkipIf(Condition.true('autest sometimes does not capture output on the last test case of a scenario'))
```

There seems to be an issue in autest with some buffered I/O. A significant precentage of the time, after tcp_client has flushed the bytes it received to stdout, autest shows failure with no output. I've observed this in Apache CI as well as my local CentOS7 environment. This happens only on the last test case in a scenario, before the Trafficserver under test is stopped. If it were a problem with the test itself, we would expect to see it happen on all test cases. So I think autest is missing a read or not flushing a write on the last time through a loop. I did not have time to look, and I would rather not enable an automated test with false failures some of the time. I used these tests to verify the correctness of the changes manually, so I think they are valuable to keep so that we can enabled them if/when we find and fix the problem.

SolidWallOfCode replied:
I think this has been found to be an AuTest issue - Jesse is tracking it down.

d2r replied:
Planning to add the skip until we can figure out what is wrong.

SolidWallOfCode commented:
[approve ci autest]

d2r commented on doc/admin-guide/files/records.config.en.rst:
```diff
@@ -1326,6 +1326,53 @@ HTTP Redirection
+   .. warning:: Following a redirect to other than ``routable`` addresses can be
+      dangerous, as it allows the controller of an origin to arrange a probe the
+      |TS| host. Generally, an administrator should never configure |TS| to
+      follow these redirects.
```
* Is this the suggested rewording?`Generally, an administrator should never configure |TS| to follow these redirects for a publicly accessible host`

* For organizations with a zero trust model, lateral attacks from within the network could still pose a risk here. I am fine with rewording this how we would like. If we do say something here I was thinking it would be good not to be overly broad in the cases when this would seem acceptable.

SolidWallOfCode replied:
Maybe "enabling these redirects makes |TS| open to third party attacks and probing and therefore should only be considered in known safe environments".

d2r replied:
Sounds good, will update it.

d2r commented on doc/admin-guide/files/records.config.en.rst:
```diff
@@ -1326,6 +1326,53 @@ HTTP Redirection
+
+   If duplicate keys are present in the setting, the right-most key-value pair
+   is used. The default value is ``routable:follow``, which means "follow
+   routable redirects, return all other redirects".
```
Good idea. Will add that you need to set a number of redirects to be greater than `0`.

d2r commented on proxy/http/HttpConfig.cc:
```diff
@@ -1603,3 +1606,137 @@ HttpConfig::parse_ports_list(char *ports_string)
+  // LOOPBACK
+  action = configMapping.find(AddressClass::LOOPBACK) != configMapping.end() ? configMapping[AddressClass::LOOPBACK] :
+                                                                               configMapping[AddressClass::DEFAULT];
+  min.load("127.0.0.0");
```
Sure we can do that.

To be clear, these would be `static const IpAddr private10_8_min;` and `static const IpAddr private10_8_max`, followed by `ret->mark(private10_8_min, private_10_8_max, reinterpret_cast<void *>(action));` ?

d2r commented on proxy/http/HttpConfig.h:
```diff
@@ -823,6 +864,10 @@ struct HttpConfigParams : public ConfigInfo {
   MgmtInt post_copy_size = 2048;
   MgmtInt max_post_size  = 0;
 
+  char *redirect_actions_string                        = nullptr;
```
This holds the string from the config file before it is parsed. It is allocated by the `HttpEstablishStaticConfigStringAlloc` macro in HttpConfig and is freed by the `HttpConfigParams` destructor.

I tried to follow the pattern used by existing configs with parsers.

d2r commented on proxy/http/HttpConfig.h:
```diff
@@ -926,8 +974,8 @@ inline HttpConfigParams::~HttpConfigParams()
   ats_free(oride.cache_vary_default_other);
   ats_free(connect_ports_string);
   ats_free(reverse_proxy_no_host_redirect);
+  ats_free(redirect_actions_string);
```
Skimming part of the call tree from the `HttpEstablishStaticConfigStringAlloc` macro, it seems using a string type might be very tricky to do. The memory here is allocated internally by some RecCore code, and this is a pointer to that. It would take some time to change it, and it would look rather odd being the only string that is allocated in a different way than the others. If we want to go down that route we can, but it would take me some time to figure it out.

d2r replied:
Won't fix this here. See [comment](https://github.com/apache/trafficserver/pull/4145/files#r216478374).

d2r commented on doc/admin-guide/files/records.config.en.rst:
```diff
@@ -1326,6 +1326,53 @@ HTTP Redirection
+   ``private``   IPv4 ``10.0.0.0/8`` ``100.64.0.0/10`` ``172.16.0.0/12`` ``192.168.0.0/16`` and IPv6 ``fc00::/7``
+   ``multicast`` IPv4 ``224.0.0.0/4`` and IPv6 ``ff00::/8``
+   ``linklocal`` IPv4 ``169.254.0.0/16`` and IPv6 ``fe80::/10``
+   ``self``      Addresses on which |TS| listens
```
Nice catch.
These are the addresses that are found using either getifaddrs(3) or ioctl(2) with `SIOCGIFCONF`, which tell you the interfaces on the local machine and not necessarily those on which ATS is listening. I'll reword this.

d2r commented on doc/admin-guide/files/records.config.en.rst:
```diff
@@ -1326,6 +1326,61 @@ HTTP Redirection
+   redirects, return all other redirects". Note that
+   :ts:cv:`proxy.config.http.number_of_redirections` must be positive also,
+   otherwise redirects will be returned rather than followed.
+
```
![screen shot 2018-09-10 at 16 53 43](https://user-images.githubusercontent.com/905298/45326767-7b776700-b51a-11e8-92c5-df78da02700b.png)

Rendering at a9e1f1ef8726e66ea35a950b59290f491721c7e1

d2r replied:
![screen shot 2018-09-11 at 13 10 48](https://user-images.githubusercontent.com/905298/45378846-2b54df00-b5c4-11e8-9e8c-4061d5306be8.png)

Rendering at 1cda6316fdf7b53b4083b64abff30d7b029fd908

bryancall commented on proxy/http/HttpConfig.cc:
```diff
@@ -1603,3 +1606,137 @@ HttpConfig::parse_ports_list(char *ports_string)
+  using RedirectEnabled::address_class_map;
+
+  if (nullptr == input_string) {
+    Emergency("parse_redirect_actions: The configuration value is empty.");
```
Emergency isn't used in the code.  I would go with Error.  That is what we do with configuration parsing.

d2r replied:
I think we originally had an `ink_release_assert` here, and @SolidWallOfCode suggested we use `Emergency` instead. What this appeared to do was indicate a non-recoverable error. I think `Error` will not do the same, but I could be mistaken. In this case the error is not recoverable.

bryancall replied:
@d2r If you want the process to die on a error I would use Fatal. It is used all over traffic server for logging Fatal errors and exiting.

Update: Talked to @d2r on slack that Emergency exists with UNRECOVERABLE_EXIT and traffic_manager will not restart the process.

bryancall commented on proxy/http/HttpTransact.cc:
```diff
@@ -1546,6 +1546,36 @@ HttpTransact::OSDNSLookup(State *s)
+      SET_VIA_STRING(VIA_DETAIL_TUNNEL, VIA_DETAIL_TUNNEL_NO_FORWARD);
+      TRANSACT_RETURN(SM_ACTION_SEND_ERROR_CACHE_NOOP, nullptr);
+      break;
+    case RedirectEnabled::Action::RETURN:
```
Why not disable redirect for the transaction and return the origins redirect response?

d2r replied:
I think this is because we are already in the process of redirecting, and so disabling redirect for the transaction here would be too late.

This is tricky because of the way the state machine transitions are set up. What we need to do is resolve the target (`Location`) of the redirect response and then check the resulting address against our configuration in order to figure out what the administrator wants to do. But we only proceed to the DNS lookup state *after* we have "committed" to following the redirect. 

* If we disable redirect for the transaction too early, we will not be able to resolve the host and we will potentially return responses we should have followed or rejected. 
* If we wait until after host resolution to disable redirect for the transaction, then the logic will have no effect, and we will potentially follow redirects that we should have returned or rejected.

bryancall replied:
@d2r If you want the process to die on a error I would use Fatal.  It is used all over traffic_server for logging Fatal errors and exiting.

d2r replied:
> @d2r If you want the process to die on a error I would use Fatal. It is used all over traffic server for logging Fatal errors and exiting.

Created #4229 as discussed so that there is clarity on why `Emergency` exists apart from `Fatal`.

d2r commented:
Squashed the branch. Old commit was 59422e07d674be135b41f0573dea475624289cc5.

d2r commented:
I am going to rebase, because an `ink_release_assert` triggered in one of my local tests that was likely due to a bug that has been fixed on master branch after the commit on which I have based:

```
(gdb) p history
$7 = {history = {{location = {file = 0x826d62 "HttpSM.cc",
        func = 0x829cc0 <HttpSM::state_read_client_request_header(int, void*)::__FUNCTION__> "state_read_client_request_header",
        line = 603}, event = 100, reentrancy = 2}, {location = {file = 0x826d62 "HttpSM.cc",
        func = 0x82a8a8 <HttpSM::set_next_state()::__FUNCTION__> "set_next_state", line = 7274}, event = 34463, reentrancy = 2}, {
      location = {file = 0x826d62 "HttpSM.cc", func = 0x82a8a8 <HttpSM::set_next_state()::__FUNCTION__> "set_next_state", line = 7320},
      event = 34463, reentrancy = 2}, {location = {file = 0x826d62 "HttpSM.cc",
        func = 0x829eb0 <HttpSM::state_http_server_open(int, void*)::__FUNCTION__> "state_http_server_open", line = 1731}, event = 200,
      reentrancy = 3}, {location = {file = 0x826d62 "HttpSM.cc",
        func = 0x829d10 <HttpSM::state_watch_for_client_abort(int, void*)::__FUNCTION__> "state_watch_for_client_abort", line = 847},
      event = 104, reentrancy = 1}, {location = {file = 0x826d62 "HttpSM.cc",
        func = 0x829eb0 <HttpSM::state_http_server_open(int, void*)::__FUNCTION__> "state_http_server_open", line = 1731}, event = 101,
      reentrancy = 1}, {location = {file = 0x826d62 "HttpSM.cc",
        func = 0x829f20 <HttpSM::state_send_server_request_header(int, void*)::__FUNCTION__> "state_send_server_request_header",
        line = 2003}, event = 103, reentrancy = 1}, {location = {file = 0x826d62 "HttpSM.cc",
        func = 0x829ee0 <HttpSM::state_read_server_response_header(int, void*)::__FUNCTION__> "state_read_server_response_header",
        line = 1845}, event = 100, reentrancy = 1}, {location = {file = 0x826d62 "HttpSM.cc",
        func = 0x82a8a8 <HttpSM::set_next_state()::__FUNCTION__> "set_next_state", line = 7274}, event = 34463, reentrancy = 1}, {
      location = {file = 0x826d62 "HttpSM.cc",
        func = 0x829fb0 <HttpSM::state_hostdb_lookup(int, void*)::__FUNCTION__> "state_hostdb_lookup", line = 2235}, event = 500,
      reentrancy = 1}, {location = {file = 0x0, func = 0x0, line = 0}, event = 0, reentrancy = 0} <repeats 55 times>}, history_pos = 10}
```

This shows a HostDB lookup event happening after we have already received a response from origin. This shouldn't happen, and it does not seem to be related to this PR's changes.

d2r commented on proxy/http/HttpConfig.h:
```diff
@@ -789,6 +824,12 @@ struct HttpConfigParams : public ConfigInfo {
+    REDIRECT_ACTIONS_REJECT = 1,
+    REDIRECT_ACTIONS_FOLLOW = 2,
+  };
+
```
These are unused. Will remove.

d2r commented:
Rebased. Old commit was 1cda6316fdf7b53b4083b64abff30d7b029fd908.

d2r commented:
We have found that there is an assert when DNS lookup fails when attempting to evaluate the redirect response. Need to fix it.

d2r commented:
```Diff
diff --git a/proxy/http/HttpTransact.cc b/proxy/http/HttpTransact.cc
index 09455e5d7..950501ef8 100644
--- a/proxy/http/HttpTransact.cc
+++ b/proxy/http/HttpTransact.cc
@@ -1546,36 +1546,6 @@ HttpTransact::OSDNSLookup(State *s)
     TRANSACT_RETURN(SM_ACTION_SEND_ERROR_CACHE_NOOP, nullptr);
   }
 
-  if (s->redirect_info.redirect_in_process) {
-    RedirectEnabled::Action action = RedirectEnabled::Action::INVALID;
-    if (true == Machine::instance()->is_self(s->host_db_info.ip())) {
-      action = s->http_config_param->redirect_actions_self_action;
-    } else {
-      ink_release_assert(s->http_config_param->redirect_actions_map != nullptr);
-      ink_release_assert(
-        s->http_config_param->redirect_actions_map->contains(s->host_db_info.ip(), reinterpret_cast<void **>(&action)));
-    }
-    switch (action) {
-    case RedirectEnabled::Action::FOLLOW:
-      TxnDebug("http_trans", "[OSDNSLookup] Invalid redirect address. Following");
-      break;
-    case RedirectEnabled::Action::REJECT:
-      TxnDebug("http_trans", "[OSDNSLookup] Invalid redirect address. Rejecting.");
-      build_error_response(s, HTTP_STATUS_FORBIDDEN, nullptr, "request#syntax_error");
-      SET_VIA_STRING(VIA_DETAIL_TUNNEL, VIA_DETAIL_TUNNEL_NO_FORWARD);
-      TRANSACT_RETURN(SM_ACTION_SEND_ERROR_CACHE_NOOP, nullptr);
-      break;
-    case RedirectEnabled::Action::RETURN:
-      TxnDebug("http_trans", "[OSDNSLookup] Configured to return on invalid redirect address.");
-      // fall-through
-    default:
-      // Return this 3xx to the client as-is.
-      TxnDebug("http_trans", "[OSDNSLookup] Invalid redirect address. Returning.");
-      build_response_copy(s, &s->hdr_info.server_response, &s->hdr_info.client_response, s->client_info.http_version);
-      TRANSACT_RETURN(SM_ACTION_INTERNAL_CACHE_NOOP, nullptr);
-    }
-  }
-
   // detect whether we are about to self loop. the client may have
   // specified the proxy as the origin server (badness).
   // Check if this procedure is already done - YTS Team, yamsat
@@ -1637,6 +1607,37 @@ HttpTransact::OSDNSLookup(State *s)
   ink_assert(s->dns_info.lookup_success);
   TxnDebug("http_seq", "[HttpTransact::OSDNSLookup] DNS Lookup successful");
 
+  if (s->redirect_info.redirect_in_process) {
+    // If dns lookup was not successful, the code below will handle the error.
+    RedirectEnabled::Action action = RedirectEnabled::Action::INVALID;
+    if (true == Machine::instance()->is_self(s->host_db_info.ip())) {
+      action = s->http_config_param->redirect_actions_self_action;
+    } else {
+      ink_release_assert(s->http_config_param->redirect_actions_map != nullptr);
+      ink_release_assert(
+        s->http_config_param->redirect_actions_map->contains(s->host_db_info.ip(), reinterpret_cast<void **>(&action)));
+    }
+    switch (action) {
+    case RedirectEnabled::Action::FOLLOW:
+      TxnDebug("http_trans", "[OSDNSLookup] Invalid redirect address. Following");
+      break;
+    case RedirectEnabled::Action::REJECT:
+      TxnDebug("http_trans", "[OSDNSLookup] Invalid redirect address. Rejecting.");
+      build_error_response(s, HTTP_STATUS_FORBIDDEN, nullptr, "request#syntax_error");
+      SET_VIA_STRING(VIA_DETAIL_TUNNEL, VIA_DETAIL_TUNNEL_NO_FORWARD);
+      TRANSACT_RETURN(SM_ACTION_SEND_ERROR_CACHE_NOOP, nullptr);
+      break;
+    case RedirectEnabled::Action::RETURN:
+      TxnDebug("http_trans", "[OSDNSLookup] Configured to return on invalid redirect address.");
+      // fall-through
+    default:
+      // Return this 3xx to the client as-is.
+      TxnDebug("http_trans", "[OSDNSLookup] Invalid redirect address. Returning.");
+      build_response_copy(s, &s->hdr_info.server_response, &s->hdr_info.client_response, s->client_info.http_version);
+      TRANSACT_RETURN(SM_ACTION_INTERNAL_CACHE_NOOP, nullptr);
+    }
+  }
+
   if (DNSLookupInfo::OS_Addr::OS_ADDR_TRY_HOSTDB == s->dns_info.os_addr_style) {
     // We've backed off from a client supplied address and found some
     // HostDB addresses. We use those if they're different from the CTA.
diff --git a/tests/gold_tests/redirect/redirect_actions.test.py b/tests/gold_tests/redirect/redirect_actions.test.py
index d679f9808..ae4e0f441 100644
--- a/tests/gold_tests/redirect/redirect_actions.test.py
+++ b/tests/gold_tests/redirect/redirect_actions.test.py
@@ -26,7 +26,7 @@ Test redirection behavior to invalid addresses
 '''
 
 Test.SkipIf(Condition.true('autest sometimes does not capture output on the last test case of a scenario'))
-Test.ContinueOnFail = True
+Test.ContinueOnFail = False
 
 Test.Setup.Copy(os.path.join(Test.Variables.AtsTestToolsDir,'tcp_client.py'))
 
@@ -82,6 +82,8 @@ def normalizeForAutest(value):
     annoying characters.
     This means we can also use them in URLs.
     '''
+    if not value:
+        return None
     return re.sub(r'[^a-z0-9-]', '_', value, flags=re.I)
 
 def makeTestCase(redirectTarget, expectedAction, scenario):
@@ -126,7 +128,8 @@ def makeTestCase(redirectTarget, expectedAction, scenario):
     testDomain = 'testdomain{0}.test'.format(normRedirectTarget)
     # The micro DNS server can't tell us whether it has a record of the domain already, so we use a dictionary to avoid duplicates.
     # We remove any surrounding brackets that are common to IPv6 addresses.
-    dnsRecords[testDomain] = [redirectTarget.strip('[]')]
+    if redirectTarget:
+        dnsRecords[testDomain] = [redirectTarget.strip('[]')]
 
     # A GET request parameterized on the config and on the target.
     request_header = {
@@ -178,6 +181,9 @@ class ActionE(Enum):
     Reject = {'config':'reject', 'expectedStatusLine':'HTTP/1.1 403 Forbidden\r\n'}
     Follow = {'config':'follow', 'expectedStatusLine':'HTTP/1.1 204 No Content\r\n'}
 
+    # Added to test failure modes.
+    Break  = {'expectedStatusLine': 'HTTP/1.1 502 Cannot find server.\r\n'}
+
 scenarios = [
     {
         # Follow to loopback, but alternately reject/return others.
@@ -228,6 +234,9 @@ for scenario in scenarios:
             expectedAction = scenario[addressClass] if addressClass in scenario else scenario[AddressE.Default]
             makeTestCase(redirectTarget=address, expectedAction=expectedAction, scenario=scenario)
 
+    # Test redirects to names that cannot be resolved.
+    makeTestCase(redirectTarget=None, expectedAction=ActionE.Break, scenario=scenario)
+
 dns.addRecords(records=dnsRecords)
 
 # Make sure this runs only after local files have been created.
```

Rebased with the fix for HostDB lookup failure and test cases that cover that failure mode.

Previous commit was 9001a5dfdcdfc7dcca1a7a5d40a8d39fe7f1dbc4.

SolidWallOfCode commented on proxy/http/HttpTransact.cc:
```diff
@@ -1607,6 +1607,37 @@ HttpTransact::OSDNSLookup(State *s)
+    }
+  }
+
   if (DNSLookupInfo::OS_Addr::OS_ADDR_TRY_HOSTDB == s->dns_info.os_addr_style) {
```
I think you might want to move your check after this, because it can change `host_db_info`.

d2r replied:
Correct. We want to use the hostdb address if we have one that is different, and not act on the address of the client-supplied host in that case.

I will fix this.

d2r commented:
Addressed the review [comment](https://github.com/apache/trafficserver/pull/4145#discussion_r217209175) by moving the logic lower in the method but leaving it unchanged.

Here is the diff:
```Diff
diff --git a/proxy/http/HttpTransact.cc b/proxy/http/HttpTransact.cc
index 950501ef8..7c0e933d1 100644
--- a/proxy/http/HttpTransact.cc
+++ b/proxy/http/HttpTransact.cc
@@ -1607,37 +1607,6 @@ HttpTransact::OSDNSLookup(State *s)
   ink_assert(s->dns_info.lookup_success);
   TxnDebug("http_seq", "[HttpTransact::OSDNSLookup] DNS Lookup successful");
 
-  if (s->redirect_info.redirect_in_process) {
-    // If dns lookup was not successful, the code below will handle the error.
-    RedirectEnabled::Action action = RedirectEnabled::Action::INVALID;
-    if (true == Machine::instance()->is_self(s->host_db_info.ip())) {
-      action = s->http_config_param->redirect_actions_self_action;
-    } else {
-      ink_release_assert(s->http_config_param->redirect_actions_map != nullptr);
-      ink_release_assert(
-        s->http_config_param->redirect_actions_map->contains(s->host_db_info.ip(), reinterpret_cast<void **>(&action)));
-    }
-    switch (action) {
-    case RedirectEnabled::Action::FOLLOW:
-      TxnDebug("http_trans", "[OSDNSLookup] Invalid redirect address. Following");
-      break;
-    case RedirectEnabled::Action::REJECT:
-      TxnDebug("http_trans", "[OSDNSLookup] Invalid redirect address. Rejecting.");
-      build_error_response(s, HTTP_STATUS_FORBIDDEN, nullptr, "request#syntax_error");
-      SET_VIA_STRING(VIA_DETAIL_TUNNEL, VIA_DETAIL_TUNNEL_NO_FORWARD);
-      TRANSACT_RETURN(SM_ACTION_SEND_ERROR_CACHE_NOOP, nullptr);
-      break;
-    case RedirectEnabled::Action::RETURN:
-      TxnDebug("http_trans", "[OSDNSLookup] Configured to return on invalid redirect address.");
-      // fall-through
-    default:
-      // Return this 3xx to the client as-is.
-      TxnDebug("http_trans", "[OSDNSLookup] Invalid redirect address. Returning.");
-      build_response_copy(s, &s->hdr_info.server_response, &s->hdr_info.client_response, s->client_info.http_version);
-      TRANSACT_RETURN(SM_ACTION_INTERNAL_CACHE_NOOP, nullptr);
-    }
-  }
-
   if (DNSLookupInfo::OS_Addr::OS_ADDR_TRY_HOSTDB == s->dns_info.os_addr_style) {
     // We've backed off from a client supplied address and found some
     // HostDB addresses. We use those if they're different from the CTA.
@@ -1678,6 +1647,37 @@ HttpTransact::OSDNSLookup(State *s)
            "IP: %s",
            ats_ip_ntop(&s->server_info.dst_addr.sa, addrbuf, sizeof(addrbuf)));
 
+  if (s->redirect_info.redirect_in_process) {
+    // If dns lookup was not successful, the code below will handle the error.
+    RedirectEnabled::Action action = RedirectEnabled::Action::INVALID;
+    if (true == Machine::instance()->is_self(s->host_db_info.ip())) {
+      action = s->http_config_param->redirect_actions_self_action;
+    } else {
+      ink_release_assert(s->http_config_param->redirect_actions_map != nullptr);
+      ink_release_assert(
+        s->http_config_param->redirect_actions_map->contains(s->host_db_info.ip(), reinterpret_cast<void **>(&action)));
+    }
+    switch (action) {
+    case RedirectEnabled::Action::FOLLOW:
+      TxnDebug("http_trans", "[OSDNSLookup] Invalid redirect address. Following");
+      break;
+    case RedirectEnabled::Action::REJECT:
+      TxnDebug("http_trans", "[OSDNSLookup] Invalid redirect address. Rejecting.");
+      build_error_response(s, HTTP_STATUS_FORBIDDEN, nullptr, "request#syntax_error");
+      SET_VIA_STRING(VIA_DETAIL_TUNNEL, VIA_DETAIL_TUNNEL_NO_FORWARD);
+      TRANSACT_RETURN(SM_ACTION_SEND_ERROR_CACHE_NOOP, nullptr);
+      break;
+    case RedirectEnabled::Action::RETURN:
+      TxnDebug("http_trans", "[OSDNSLookup] Configured to return on invalid redirect address.");
+      // fall-through
+    default:
+      // Return this 3xx to the client as-is.
+      TxnDebug("http_trans", "[OSDNSLookup] Invalid redirect address. Returning.");
+      build_response_copy(s, &s->hdr_info.server_response, &s->hdr_info.client_response, s->client_info.http_version);
+      TRANSACT_RETURN(SM_ACTION_INTERNAL_CACHE_NOOP, nullptr);
+    }
+  }
+
   // so the dns lookup was a success, but the lookup succeeded on
   // a hostname which was expanded by the traffic server. we should
   // not automatically forward the request to this expanded hostname.
```

Branch has been squashed. Previous commit was bb21f28088861bbf02856514b4d2dfdeb6e921cd.
